aten/src/ATen/native/quantized/cpu/qnnpack/src/hardswish.c

```diff
@@ -0,0 +1,164 @@
+#include <math.h>
+#include <stddef.h>
+#include <stdint.h>
+#include <stdlib.h>
+
+#include <pytorch_qnnpack.h>
+#include <qnnpack/log.h>
+#include <qnnpack/operator.h>
+
+enum pytorch_qnnp_status pytorch_qnnp_create_hardswish_nc_q8(
+    size_t channels,
+    uint8_t input_zero_point,
+    float input_scale,
+    uint8_t output_zero_point,
+    float output_scale,
+    uint8_t output_min,
+    uint8_t output_max,
+    uint32_t flags,
+    pytorch_qnnp_operator_t* hardswish_out) {
+  pytorch_qnnp_operator_t hardswish_op = NULL;
+  enum pytorch_qnnp_status status = pytorch_qnnp_status_uninitialized;
+
+  if (!pytorch_qnnp_params.initialized) {
+    pytorch_qnnp_log_error(
+        "pytorch_qnnp_create_hardswish_nc_q8 failed because QNNPACK is not properly initialized");
+    goto error;
+  }
+
+  status = pytorch_qnnp_status_invalid_parameter;
+
+  if (channels == 0) {
+    pytorch_qnnp_log_error(
+        "failed to create Hardswish operator with %zu channels: number of channels must be non-zero",
+        channels);
+    goto error;
+  }
+
+  if (input_scale <= 0.0f || !isnormal(input_scale)) {
+    pytorch_qnnp_log_error(
+        "failed to create Hardswish operator with %.7g input scale: scale must be finite and positive",
+        input_scale);
+    goto error;
+  }
+
+  if (output_scale <= 0.0f || !isnormal(output_scale)) {
+    pytorch_qnnp_log_error(
+        "failed to create Hardswish operator with %.7g output scale: scale must be finite and positive",
+        output_scale);
+    goto error;
+  }
+
+  if (output_min >= output_max) {
+    pytorch_qnnp_log_error(
+        "failed to create Hardswish operator with [%" PRIu8 ", %" PRIu8
+        "] output range: range min must be below range max",
+        output_min,
+        output_max);
+    goto error;
+  }
+
+  status = pytorch_qnnp_status_unsupported_parameter;
+
+  if (output_scale != input_scale) {
+    pytorch_qnnp_log_error(
+        "failed to create Hardswish operator with %.7g output scale: only output scale equal to input scale is supported",
+        output_scale);
+    goto error;
+  }
+
+  if (output_zero_point != input_zero_point) {
+    pytorch_qnnp_log_error(
+        "failed to create Hardswish operator with %" PRIu8
+        " output zero point: only output zero point equal to input zero point is supported",
+        output_zero_point);
+    goto error;
+  }
+
+  status = pytorch_qnnp_status_out_of_memory;
+
+  hardswish_op = calloc(1, sizeof(struct pytorch_qnnp_operator));
+  if (hardswish_op == NULL) {
+    pytorch_qnnp_log_error(
+        "failed to allocate %zu bytes for pytorch_qnnp_operator structure",
+        sizeof(struct pytorch_qnnp_operator));
+    goto error;
+  }
+
+  hardswish_op->lookup_table = malloc(256 * sizeof(uint8_t));
+  if (hardswish_op->lookup_table == NULL) {
+    pytorch_qnnp_log_error(
+        "failed to allocate 256 bytes for Hardswish lookup table");
+    goto error;
+  }
+
+  uint8_t* lookup_table = hardswish_op->lookup_table;
```

Comment:
I wonder if there is a requirement or interest in supporting other data types for the LUT. But that would involve re-writing the kernel, I think. Might be good to mark it as a TODO (if server supports it)

Reply:
from what I understand uint8_t is the only dtype supported in QNNPACK as a whole

Reply:
@vkuzo, btw another thought about avoiding to build lookup table: You can have a model level parameter or attribute, that stores all the lookup tables that caches on quantization parameters. So now you have `quantized::lookup_lookup_table` op that given the parameters returns lookup table either wrapped in a Tensor (via cpp_type_hack) or via custom class (via torchbind). This op takes in lookup_table as input and check if the lookup table exist, else it builds one. Not too different from current conv_prepack.